internal/services/azure_reservation_service.go

```diff
@@ -0,0 +1,71 @@
+package services
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/RHEnVision/provisioning-backend/internal/clients"
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/dao"
+	"github.com/RHEnVision/provisioning-backend/internal/models"
+	"github.com/RHEnVision/provisioning-backend/internal/payloads"
+	"github.com/go-chi/render"
+)
+
+func CreateAzureReservation(w http.ResponseWriter, r *http.Request) {
+	logger := *ctxval.Logger(r.Context())
+
+	payload := &payloads.AzureReservationRequestPayload{}
+	if err := render.Bind(r, payload); err != nil {
+		renderError(w, r, payloads.NewInvalidRequestError(r.Context(), "Azure reservation", err))
+		return
+	}
+
+	pkDao := dao.GetPubkeyDao(r.Context())
+
+	// validate pubkey
+	logger.Debug().Msgf("Validating existence of pubkey %d for this account", payload.PubkeyID)
+	pk, err := pkDao.GetById(r.Context(), payload.PubkeyID)
+	if err != nil {
+		message := fmt.Sprintf("get pubkey with id %d", payload.PubkeyID)
+		renderNotFoundOrDAOError(w, r, err, message)
+		return
+	}
+	logger.Debug().Msgf("Found pubkey %d named '%s'", pk.ID, pk.Name)
+
+	// Get Sources client
+	sourcesClient, err := clients.GetSourcesClient(r.Context())
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	// Fetch SubscriptionID from Sources
+	authentication, err := sourcesClient.GetAuthentication(r.Context(), payload.SourceID)
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	if typeErr := authentication.MustBe(models.ProviderTypeAzure); typeErr != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), typeErr))
+		return
+	}
+
+	azureClient, err := clients.GetAzureClient(r.Context(), authentication)
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	var azureImageID string
+	if strings.HasPrefix(payload.ImageID, "/subscriptions/") {
+		// Direct AMI ID was provided, no need to call image builder
+		azureImageID = payload.ImageID
+	} else {
+		// TODO fetch image ID from Image Builder
+	}
+
+	azureClient.CreateVM(r.Context(), azureImageID, pk.Body)
+}
```

Comment:
Well that is ugly, the resource is reservation but it launches it directly. I guess we need the job for the first version :-)

Reply:
Sure, I get that. My point is do we want to merge this as is or refactor this to a background job for the first version? I lean towards the latter.